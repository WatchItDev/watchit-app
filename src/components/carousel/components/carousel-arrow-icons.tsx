import Iconify from '@src/components/iconify';
import { ArrowIconProps } from '@src/components/carousel/types.ts';

// ----------------------------------------------------------------------

export function LeftIcon({ icon = 'eva:arrow-ios-forward-fill', isRTL }: Readonly<ArrowIconProps>) {
  return (
    <Iconify
      data-testid="leftButton"
      icon={icon}
      sx={{
        transform: ' scaleX(-1)',
        ...(isRTL && {
          transform: ' scaleX(1)',
        }),
      }}
    />
  );
}

export function RightIcon({ icon = 'eva:arrow-ios-forward-fill', isRTL }: Readonly<ArrowIconProps>) {
  return (
    <Iconify
      data-testid="rightButton"
      icon={icon}
      sx={{
        ...(isRTL && {
          transform: ' scaleX(-1)',
        }),
      }}
    />
  );
}
