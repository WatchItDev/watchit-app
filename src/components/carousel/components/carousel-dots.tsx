import React from 'react';
import Stack from '@mui/material/Stack';
import Box from '@mui/material/Box';
import { styled } from '@mui/material/styles';
// Local Imports
import { CarouselDotsStyledRootProps, CarouselDotsProps } from '@src/components/carousel/types.ts';

// ----------------------------------------------------------------------
const StyledRoot = styled(Box, {
  shouldForwardProp: (prop) => prop !== 'rounded',
})<CarouselDotsStyledRootProps>(({ rounded, theme }) => ({
  zIndex: 9,
  margin: 0,
  padding: 0,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  color: theme.palette.primary.main,
  '& li': {
    width: 18,
    height: 18,
    opacity: 0.32,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    cursor: 'pointer',
    '&.slick-active': {
      opacity: 1,
      ...(rounded && {
        '& span': {
          width: 16,
          borderRadius: 6,
        },
      }),
    },
  },
}));

const StyledDot = styled('span')(({ theme }) => ({
  width: 8,
  height: 8,
  borderRadius: '50%',
  transition: theme.transitions.create(['width'], {
    easing: theme.transitions.easing.sharp,
    duration: theme.transitions.duration.short,
  }),
}));

// ----------------------------------------------------------------------

export default function CarouselDots(props?: CarouselDotsProps) {
  const rounded = props?.rounded || false;
  const sx = props?.sx;

  return {
    appendDots: (dots: React.ReactNode) => (
        <StyledRoot component="ul" rounded={rounded} sx={{ ...sx }} {...props}>
          {dots}
        </StyledRoot>
    ),
    customPaging: () => (
      <Stack
        component="div"
        alignItems="center"
        justifyContent="center"
        sx={{ width: 1, height: 1 }}
      >
        <StyledDot
          sx={{
            bgcolor: 'currentColor',
          }}
        />
      </Stack>
    ),
  };
}
