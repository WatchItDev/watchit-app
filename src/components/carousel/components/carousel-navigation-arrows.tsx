import { FC } from 'react';
import { icons } from '@tabler/icons-react';
import IconButton from '@mui/material/IconButton';
import { NavigationArrowsProps } from '@src/components/carousel/types.ts';

const CarouselNavigationArrows: FC<NavigationArrowsProps> = ({
  next,
  prev,
}) => (
  <>
    <IconButton data-testid={'leftButton'} onClick={prev}>
      <icons.IconChevronLeft />
    </IconButton>
    <IconButton data-testid={'rightButton'} onClick={next}>
      <icons.IconChevronRight />
    </IconButton>
  </>
);

export default CarouselNavigationArrows;
