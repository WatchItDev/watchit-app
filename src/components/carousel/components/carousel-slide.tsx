import Box from '@mui/material/Box';
import { memo, FC } from 'react';
import { CarouselSlideProps } from '../types';
// @ts-expect-error No error in this context
import { Post } from "@lens-protocol/api-bindings/dist/declarations/src/lens/graphql/generated";
import { Profile } from '@lens-protocol/api-bindings';

const CarouselSlide: FC<Readonly<CarouselSlideProps<Profile | Post>>> = (props) => {
  const { items, itemsPerRow, renderItem, } = props;
  const row1 = items.slice(0, itemsPerRow);
  const row2 = items.slice(itemsPerRow, itemsPerRow * 2);
  const itemWidthPercent = 100 / itemsPerRow;

  return (
    <Box>
      {[row1, row2].map((rowItems, _rowIndex) => (
        <Box key={`loop-one-${rowItems[_rowIndex]?.id}`} sx={{ display: 'flex' }}>
          {rowItems.map((item, _item) => (
            <Box
              key={`loop-two-${item?.id}`}
              sx={{
                flexBasis: `${itemWidthPercent}%`,
                maxWidth: `${itemWidthPercent}%`,
                p: 1,
              }}
            >
              {renderItem(item)}
            </Box>
          ))}
        </Box>
      ))}
    </Box>
  );
}

export default memo(CarouselSlide, (prevProps, nextProps) => prevProps.items === nextProps.items);
