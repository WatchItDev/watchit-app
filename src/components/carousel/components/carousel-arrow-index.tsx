import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
import { styled } from '@mui/material/styles';
import { bgBlur } from '@src/theme/css.ts';
import { CarouselArrowIndexProps } from '@src/components/carousel/types.ts';
import { useIsRTL } from '@src/hooks/components/use-rtl.ts';
import { LeftIcon, RightIcon } from './carousel-arrow-icons.tsx';

const StyledRoot = styled(Box)(({ theme }) => ({
  ...bgBlur({
    opacity: 0.48,
    color: theme.palette.grey[900],
  }),
  zIndex: 9,
  display: 'inline-flex',
  alignItems: 'center',
  position: 'absolute',
  bottom: theme.spacing(2),
  right: theme.spacing(2),
  padding: theme.spacing(0.25),
  color: theme.palette.common.white,
  borderRadius: theme.shape.borderRadius,
}));

const StyledIconButton = styled(IconButton)({
  width: 28,
  height: 28,
  padding: 0,
  opacity: 0.48,
  '&:hover': { opacity: 1 },
});

// ----------------------------------------------------------------------

export default function CarouselArrowIndex({
  index,
  total,
  onNext,
  onPrev,
  icon,
  sx,
  ...other
}: CarouselArrowIndexProps) {
  const isRTL = useIsRTL();

  return (
    <StyledRoot sx={sx} {...other}>
      <StyledIconButton color="inherit" onClick={onPrev}>
        <LeftIcon icon={icon} isRTL={isRTL} />
      </StyledIconButton>

      <Typography variant="subtitle2" component="span" sx={{ mx: 0.25 }}>
        {index + 1}/{total}
      </Typography>

      <StyledIconButton color="inherit" onClick={onNext}>
        <RightIcon icon={icon} isRTL={isRTL} />
      </StyledIconButton>
    </StyledRoot>
  );
}
