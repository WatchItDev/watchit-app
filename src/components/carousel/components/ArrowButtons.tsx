import React from 'react';
import { StyledIconButton } from './StyledIconButton';
import { LeftIcon, RightIcon } from './CarouselArrowIcons';
import { useIsRTL } from '@src/hooks/components/use-rtl';

interface ArrowButtonProps {
  onClick?: () => void;
  direction: 'left' | 'right';
  icon?: React.ReactNode;
  filled?: boolean;
  shape?: 'circular' | 'rounded';
  sx?: object;
}

export const ArrowButton: React.FC<ArrowButtonProps> = ({
                                                          onClick,
                                                          direction,
                                                          filled = false,
                                                          shape = 'circular',
                                                          icon,
                                                          sx,
                                                        }) => {
  const isRTL = useIsRTL();
  const IconComponent = direction === 'left' ? LeftIcon : RightIcon;

  return (
    <StyledIconButton filled={filled} shape={shape} onClick={onClick} sx={sx}>
      <IconComponent icon={icon} isRTL={isRTL} />
    </StyledIconButton>
  );
};
