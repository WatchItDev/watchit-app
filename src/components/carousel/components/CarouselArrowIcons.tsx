import Iconify from '@src/components/iconify';
import { ArrowIconProps } from '@src/components/carousel/components/types';

// ----------------------------------------------------------------------

export function LeftIcon({ icon = 'eva:arrow-ios-forward-fill', isRTL }: ArrowIconProps) {
  return (
    <Iconify
      icon={icon}
      sx={{
        transform: ' scaleX(-1)',
        ...(isRTL && {
          transform: ' scaleX(1)',
        }),
      }}
    />
  );
}

export function RightIcon({ icon = 'eva:arrow-ios-forward-fill', isRTL }: ArrowIconProps) {
  return (
    <Iconify
      icon={icon}
      sx={{
        ...(isRTL && {
          transform: ' scaleX(-1)',
        }),
      }}
    />
  );
}
