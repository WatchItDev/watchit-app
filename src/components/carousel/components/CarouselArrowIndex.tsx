import { Typography } from '@mui/material';
import { ArrowButton } from './ArrowButtons';
import { StyledRoot } from './StyledRoot';
import {CarouselArrowIndexProps} from "@src/components/carousel/types.ts";

export default function CarouselArrowIndex({
                                             index,
                                             total,
                                             onNext,
                                             onPrev,
                                             icon,
                                             sx,
                                             ...other
                                           }: CarouselArrowIndexProps) {
  return (
    <StyledRoot blur sx={sx} {...other}>
      <ArrowButton direction="left" onClick={onPrev} icon={icon} />
      <Typography variant="subtitle2" component="span" sx={{ mx: 0.25 }}>
        {index + 1}/{total}
      </Typography>
      <ArrowButton direction="right" onClick={onNext} icon={icon} />
    </StyledRoot>
  );
}
