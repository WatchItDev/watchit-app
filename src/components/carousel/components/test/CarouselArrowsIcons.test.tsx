import { it, describe, expect } from 'vitest';
import { render } from '@testing-library/react';
import {LeftIcon, RightIcon} from '@src/components/carousel/components/CarouselArrowIcons';

describe('LeftIcon component', () => {
  it('to match snapshot', () => {
    const { baseElement } = render(<LeftIcon icon="eva:arrow-ios-forward-fill" isRTL={false} />);
    expect(baseElement).toMatchSnapshot();
  })

  it('renders correctly with isRTL false', () => {
    const {getByTestId} = render(<LeftIcon icon="eva:arrow-ios-forward-fill" isRTL={false} />);
    const iconElement = getByTestId('iconify');
    // When isRTL is false, the default transform should be applied.
    expect(iconElement.getAttribute('data-sx')).toContain('"transform":" scaleX(-1)"');
    // and the icon prop is passed correctly.
    expect(iconElement.getAttribute('data-icon')).toBe('eva:arrow-ios-forward-fill');
  });

  it('renders correctly with isRTL true', () => {
    expect(render(<LeftIcon icon="eva:arrow-ios-forward-fill" isRTL={true} />).getByTestId(/iconify/i).getAttribute('data-sx')).toContain('"transform":" scaleX(1)"');
  });
});

describe('RightIcon component', () => {
  it('to match snapshot', () => {
    expect(render(<RightIcon icon="eva:arrow-ios-forward-fill" isRTL={false} />).baseElement).toMatchSnapshot();
  })

  it('renders correctly with isRTL false', () => {
    const {getByTestId} = render(<RightIcon icon="eva:arrow-ios-forward-fill" isRTL={false} />);
    const iconElement = getByTestId('iconify');
    // When isRTL is false, no transform should be applied.
    // The sx prop should be an empty object.
    expect(iconElement.getAttribute('data-sx')).toBe('{}');
    expect(iconElement.getAttribute('data-icon')).toBe('eva:arrow-ios-forward-fill');
  });

  it('renders correctly with isRTL true', () => {
    expect(render(<RightIcon icon="eva:arrow-ios-forward-fill" isRTL={true} />).getByTestId(/iconify/i).getAttribute('data-sx')).toContain('"transform":" scaleX(-1)"');
  });
});
