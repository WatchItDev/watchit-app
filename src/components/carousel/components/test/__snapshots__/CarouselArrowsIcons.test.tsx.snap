// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[COMPONENTS]: LeftIcon component > to match snapshot 1`] = `
<body>
  <div>
    <span
      data-icon="eva:arrow-ios-forward-fill"
      data-sx="{"transform":" scaleX(-1)"}"
      data-testid="iconify"
    />
  </div>
</body>
`;

exports[`RightIcon component > to match snapshot 1`] = `
<body>
  <div>
    <span
      data-icon="eva:arrow-ios-forward-fill"
      data-sx="{}"
      data-testid="iconify"
    />
  </div>
</body>
`;
